python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+        -------
+        self : object
+            The model
+        """
+        self._test_inputs(X,groups)
+        td = self.training_options
+        N, p = X.shape
+        self.p = p
+        self.n_groups = len(np.unique(groups))
+        rng = np.random.default_rng(int(seed))
+
+        W_, sigmals_ = self._initialize_variables(X)
+
+        X = self._transform_training_data(X)[0]
+
+        self.groups = groups
+        list_constants = []
+        for i in range(self.n_groups):
+            torch_array = torch.zeros(self.p)
+            torch_array[groups == i] = 1
+            list_constants.append(torch_array)
+
+        trainable_variables = [W_] + sigmals_
+
+        self._initialize_save_losses()
+
+        optimizer = torch.optim.SGD(
+            trainable_variables, lr=td["learning_rate"], momentum=td["momentum"]
+        )
+        softplus = nn.Softplus()
+
+        _prior = self._create_prior()
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(td["n_iterations"]):
+            idx = rng.choice(X.shape[0], size=td["batch_size"], replace=False)
+            X_batch = X[idx]
+
+            list_covs = [
+                    softplus(sigmals_[k]) * list_constants[k]
+                    for k in range(self.n_groups)
+            ]
+
+            sigma = torch.sum(list_covs,
+                dim=0,
+            )
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + torch.diag(sigma)
+
+            m = MultivariateNormal(torch.zeros(p), Sigma)
+
+            like_tot = torch.mean(m.log_prob(X_batch))
+            like_prior = _prior(trainable_variables)
+            posterior = like_tot + like_prior / N
+            loss = -1 * posterior
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            self._save_losses(i, like_tot, like_prior, posterior)
+
+        self._store_instance_variables(trainable_variables)
+
+        return self
+
+    def get_covariance(self):
+        """
+        Gets the covariance matrix
+
+        Sigma = W^TW + sigma2*I
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        covariance : np.array-like(p,p)
+            The covariance matrix
+        """
+        covariance = np.dot(self.W_.T, self.W_) + np.diag(self.sigmas_)
+        return covariance
+
+    def _create_prior(self):
+        """
+        This creates the function representing prior on pararmeters
+
+        Parameters
+        ----------
+        log_prior : function
+            The function representing the negative log density of the prior
+        """
+        pd = self.prior_options
+
+        def log_prior(trainable_variables):
+            list_gamma_log_probs = []
+            for k in self.n_groups:
+                list_gamma_log_probs.append(
+                    Gamma(pd["alpha"], pd["beta"]).log_prob(
+                        trainable_variables[k + 1]
```

Comment:
this is an example of an occasion where it would be really helpful to have the instance variables all in one place in the init.  I wanted to understand why we're indexing by `k + 1` here and had to search through the whole class definition in order to see why.  A comment here noting that the sigmas start at position one would help to calm the reader's anxiety about an off-by-one error here.